File: CinemaReservations/Java/CinemaAllocationsDomain/src/main/java/com/sdd/cinemaallocations/Row.java
package com.sdd.cinemaallocations;

import lombok.EqualsAndHashCode;
import lombok.Getter;

import java.util.ArrayList;
import java.util.List;

@EqualsAndHashCode
@Getter
public class Row {
    private String name;
    private List<Seat> seats;

    public Row(String name, List<Seat> seats) {
        this.name = name;
        this.seats = seats;
    }

    public SeatsAllocated allocateSeats(AllocateSeats allocateSeats) {
        SeatAllocation allocation = new SeatAllocation(allocateSeats.partyRequested());

        for (Seat seat : this.seats()) {

            if (seat.isAvailable())
            {
                allocation.addSeat(seat);

                if(allocation.isFulfilled()) {
                    return new SeatsAllocated(allocateSeats.partyRequested(), allocation.allocatedSeats());
                }
            }
        }
        return new NoPossibleAllocationsFound(allocateSeats.partyRequested(), new ArrayList<>());
    }

    public Row makeSeatsReserved(List<Seat> updatedSeats) {

        for (Seat newSeat : updatedSeats) {
            this.seats.set(seats.indexOf(newSeat), newSeat.reserveSeats());
        }
        return new Row(this.name, this.seats);
    }
}

File: CinemaReservations/CSharp/CinemaAllocations.Domain/Row.cs
using System.Collections.Generic;
using Value;

namespace CinemaAllocations.Domain
{
    public class Row : ValueType<Row>
    {
        public string Name { get; }
        public List<Seat> Seats { get; }

        public Row(string name, List<Seat> seats)
        {
            Name = name;
            Seats = seats;
        }
        
        public SeatsAllocated AllocateSeats(AllocateSeats allocateSeats) 
        {
            var allocation = new SeatAllocation(allocateSeats.PartyRequested);

            foreach (var seat in Seats) {

                if (seat.IsAvailable())
                {
                    allocation.AddSeat(seat);

                    if(allocation.IsFulfilled) {
                        return new SeatsAllocated(allocation.AllocatedSeats, allocateSeats.PartyRequested);
                    }
                }
            }
            return new NoPossibleAllocationsFound(allocateSeats.PartyRequested);
        }

        public Row MakeSeatsReserved(List<Seat> updatedSeats) 
        {
            foreach (var newSeat in updatedSeats)
            {
                Seats[Seats.FindIndex(s => s.Equals(newSeat))] = newSeat.ReserveSeats();
            }
            return new Row(Name, Seats);
        }

        protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
        {
            return new object[] { Name, new ListByValue<Seat>(new List<Seat>(Seats)) };
        }
    }

}